src/stirling/bpf_tools/rr/rr.cc

```diff
@@ -0,0 +1,372 @@
+                                      void const* const key, const uint32_t val_size, void* value) {
+  if (PlabackComplete()) {
+    return error::Internal("Playback complete.");
+  }
+
+  const auto event_wrapper = events_proto_.event(playback_event_idx_);
+  if (!event_wrapper.has_map_get_value_event()) {
+    return error::Internal("Map event not available.");
+  }
+
+  const auto event = event_wrapper.map_get_value_event();
+
+  if (name != event.name()) {
+    return error::Internal("Mismatched name.");
+  }
+  if (key_size != event.key().size()) {
+    return error::Internal("Mismatched key size.");
+  }
+  if (val_size != event.value().size()) {
+    return error::Internal("Mismatched data size.");
+  }
+  if (0 != memcmp(key, event.key().data(), key_size)) {
+    return error::Internal("Mismatched key.");
+  }
+
+  memcpy(value, event.value().data(), val_size);
+  ++playback_event_idx_;
+
+  return Status::OK();
+}
+
+Status BPFReplayer::ReplayMapGetKeyAndValue(const std::string& name, const uint32_t key_size,
+                                            void* key, const uint32_t val_size, void* val) {
+  if (PlabackComplete()) {
+    return error::Internal("Playback complete.");
+  }
+
+  const auto event_wrapper = events_proto_.event(playback_event_idx_);
+  if (!event_wrapper.has_map_get_value_event()) {
+    return error::Internal("Map event not available.");
+  }
+
+  const auto event = event_wrapper.map_get_value_event();
+
+  if (name != event.name()) {
+    return error::Internal("Mismatched name.");
+  }
+  if (key_size != event.key().size()) {
+    return error::Internal("Mismatched key size.");
+  }
+  if (val_size != event.value().size()) {
+    return error::Internal("Mismatched data size.");
+  }
+  memcpy(key, event.key().data(), key_size);
+  memcpy(val, event.value().data(), val_size);
+  ++playback_event_idx_;
+
+  return Status::OK();
+}
+
+StatusOr<int32_t> BPFReplayer::ReplayBPFMapCapacityEvent(const std::string& name) {
+  if (PlabackComplete()) {
+    return error::Internal("Playback complete.");
+  }
+
+  const auto event_wrapper = events_proto_.event(playback_event_idx_);
+  if (!event_wrapper.has_map_capacity_event()) {
+    return error::Internal("Map event not available.");
+  }
+
+  const auto event = event_wrapper.map_capacity_event();
+
+  if (name != event.name()) {
+    return error::Internal("Mismatched name.");
+  }
+  ++playback_event_idx_;
+
+  return event.capacity();
+}
+
+StatusOr<int32_t> BPFReplayer::ReplayBPFMapGetTableOfflineEvent(const std::string& name) {
+  if (PlabackComplete()) {
+    return error::Internal("Playback complete.");
+  }
+
+  const auto event_wrapper = events_proto_.event(playback_event_idx_);
+  if (!event_wrapper.has_map_get_table_offline_event()) {
+    return error::Internal("Map event not available.");
+  }
+
+  const auto event = event_wrapper.map_get_table_offline_event();
+
+  if (name != event.name()) {
+    return error::Internal("Mismatched name.");
```

Comment:
Having the expected vs actual name in the error string would be helpful for when this does fail.